### PR TITLE
SF-1923 Use bottom sheet for commenters to add new notes

### PR DESCRIPTION
**Pull request by nigel-wells**
- Adding notes via the mobile bottom sheet now displays a textarea
- Users can now add notes via the bottom sheet form
- Refactored notes to be added via the editor component rather than the note dialog

image

- - -
This change is [Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1765)

### CODE REVIEW

**Review by nigel-wells — COMMENTED** [review]
Well spotted. I've changed it to dismiss the bottom sheet while leaving the segment still selected to keep it consistent.

> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1765#-:-NRem_OY8yS5FeZWbR3B:b-c6mc4l)* status: 3 of 7 files reviewed, 4 unresolved discussions (waiting on @RaymondLuong3)

___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts` line 887 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/1765#-NRcmCTw7B4UJleosjdu:-NRei_WTEIHE8NuOiBuu:b-896fix) ([raw file](https://github.com/sillsdev/web-xforge/blob/a36b5e3707cc8ed707e2fef357357735b3ea5aac/src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts#L887)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

Add a type here
</blockquote></details>

Done.
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts` line 923 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/1765#-NRcnY1WBvAQqjZRDaEK:-NReikoh5rtyu2DLrIvX:b-rig91v) ([raw file](https://github.com/sillsdev/web-xforge/blob/a36b5e3707cc8ed707e2fef357357735b3ea5aac/src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts#L923)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

It might be good here to use the `getNoteThreadDocId` function.
</blockquote></details>

Oh yeah, I think I even referred to that in another PR recently lol. Done
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts` line 943 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/1765#-NRcn2ETCr_hC6v23RUL:-NRejW0C8Ute7VRZdD2g:b-896fix) ([raw file](https://github.com/sillsdev/web-xforge/blob/a36b5e3707cc8ed707e2fef357357735b3ea5aac/src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts#L943)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

Can you add a comment here for why the 100ms delay?
</blockquote></details>

Done.
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.spec.ts` line 3357 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/1765#-NRcrr9S1vHo-Fyhgw8e:-NRejfSX2fTh4oOqydrT:b-896fix) ([raw file](https://github.com/sillsdev/web-xforge/blob/a36b5e3707cc8ed707e2fef357357735b3ea5aac/src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.spec.ts#L3357)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

A better name would be saveMobileNoteButton
</blockquote></details>

Done.

**Review by RaymondLuong3 — CHANGES_REQUESTED** [review]
> Reviewed 4 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1765)* status: all files reviewed, 4 unresolved discussions (waiting on @nigel-wells)

___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts` line 925 at r3](https://reviewable.io/reviews/sillsdev/web-xforge/1765#-NRi1WoC0v_Pmx-fS1xp:-NRi1WoC0v_Pmx-fS1xq:ba4x427) ([raw file](https://github.com/sillsdev/web-xforge/blob/0f706647e5c018b25a7a1acef81a5173edbf1b38/src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts#L925)):*
> ```XML
>       };
>       await this.projectService.createNoteThread(this.projectId, noteThread);
>       await this.updateNoteReadRefs(note.dataId);
> ```

I merged in a change so that notes created by the current user do not need to be added to the notesReadRefs property. So this line is not needed anymore.



_Code quote:_
```XML
    await this.updateNoteReadRefs(note.dataId);
```
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.spec.ts` line 2672 at r3](https://reviewable.io/reviews/sillsdev/web-xforge/1765#-NRi21cF8CAxfxEDRYpk:-NRi21cF8CAxfxEDRYpl:b3n6wm) ([raw file](https://github.com/sillsdev/web-xforge/blob/0f706647e5c018b25a7a1acef81a5173edbf1b38/src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.spec.ts#L2672)):*
> ```XML
>       expect(noteThread.notes[0].content).toEqual(content);
>       expect(noteThread.notes[0].tagId).toEqual(2);
>       const projectUserConfigDoc: SFProjectUserConfigDoc = env.getProjectUserConfigDoc(userId);
> ```

This test lines are unnecessary now since I introduced SF-1878



_Code quote:_
```XML
      const projectUserConfigDoc: SFProjectUserConfigDoc = env.getProjectUserConfigDoc(userId);
      expect(projectUserConfigDoc.data!.noteRefsRead).toContain(noteThread.notes[0].dataId);
```

**Review by nigel-wells — COMMENTED** [review]
Thanks for the heads up on the recent changes. There were a number of conflicts during the rebase, can you do a sanity check to ensure everything you had changed is still working as I did discard a bunch of code due to the nature of this PR moving it from the dialog to the editor component.

> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1765#-:-NS3p09X5nhvTSmE2UAa:b-3vze42)* status: 5 of 7 files reviewed, 2 unresolved discussions (waiting on @RaymondLuong3)

___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts` line 925 at r3](https://reviewable.io/reviews/sillsdev/web-xforge/1765#-NRi1WoC0v_Pmx-fS1xp:-NS3ESdQDdII8fkfyUj4:b-896fix) ([raw file](https://github.com/sillsdev/web-xforge/blob/0f706647e5c018b25a7a1acef81a5173edbf1b38/src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts#L925)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

I merged in a change so that notes created by the current user do not need to be added to the notesReadRefs property. So this line is not needed anymore.
</blockquote></details>

Done.
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.spec.ts` line 2672 at r3](https://reviewable.io/reviews/sillsdev/web-xforge/1765#-NRi21cF8CAxfxEDRYpk:-NS3p-Dp1LLeEPQ7Lu1b:b-896fix) ([raw file](https://github.com/sillsdev/web-xforge/blob/0f706647e5c018b25a7a1acef81a5173edbf1b38/src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.spec.ts#L2672)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

This test lines are unnecessary now since I introduced SF-1878
</blockquote></details>

Done.

**Review by RaymondLuong3 — COMMENTED** [review]
> Reviewed 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1765)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @nigel-wells)

**Review by RaymondLuong3 — APPROVED** [review]
Thanks for working on the merge conflicts. I took a closer look, and it looks like your changes are all there.

> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1765#-:-NS7DKg70-uvKybKwrby:b7b3h6b)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @nigel-wells)

**Review by RaymondLuong3 — COMMENTED** [review]
There were some concerns about the reviewer note label, so @Nateowami temporarily removed that logic. Sorry for the confusion.

> Reviewed 4 of 5 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1765#-:-NSl85JJ5LrHuFjPzyPP:bn72nx5)* status: 6 of 7 files reviewed, all discussions resolved (waiting on @nigel-wells)

**Review by nigel-wells — COMMENTED** [review]
Ok, I've rebased again to pull in those changes to remove the reviewer note label.

> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1765#-:-NSmhpvzAywPxa4ZkJf8:bwohw5m)* status: 5 of 7 files reviewed, all discussions resolved (waiting on @RaymondLuong3)

**Review by RaymondLuong3 — APPROVED** [review]
It looks good. Did you get the latest comment from Arun about the issue he was seeing?

> Reviewed 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1765#-:-NSvZEI00ahZb1-OTSee:bd7tsyb)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @nigel-wells)

File: src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts [hunk 1]
```diff
@@ -849,6 +878,94 @@ export class EditorComponent extends DataLoadingComponent implements OnDestroy,
...
+    this.toggleNoteThreadVerses(true);
+  }
+
+  toggleAddingMobileNote(): void {
+    this.addingMobileNote = !this.addingMobileNote;
+    if (this.addingMobileNote) {
+      this.mobileNoteControl.reset();
+      // On-screen keyboards appearing can interfere with the resize height logic which then changes the focus
+      // Waiting for 100ms was found to be a good amount of time for that to be resolved
+      setTimeout(() => this.mobileNoteTextarea?.nativeElement.focus(), 650);
```
Comment (Nateowami) [inline src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts:962]: Are there not events we can subscribe to avoid this?

**Comment by Nateowami** [comment]
I'm seeing several issues. I don't know if you intend to deal with them in this PR, so it's perfectly fine (maybe even preferable) to put them off.

1. After adding a note, it focuses the first segment:


https://user-images.githubusercontent.com/6140710/232876553-a13ed6df-74d9-48f9-933e-ec8d0f3e90c6.mp4

2. The bottom sheet never closes itself, even if I navigate away to community checking.

**Review by nigel-wells — COMMENTED** [review]
Are you on the latest version of the branch as the bottom sheet overlaying the menu was fixed (or supposed to be) via another PR.

Your video is also not the experience I see - perhaps it is the same issue above where you're not on the up-to-date branch?

I'm keen to get this branch in somewhat soon as it is a big change from the note dialog and every time that component is touched it causes a lot of conflicts.

> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1765#-:-NTLmqR29a-UTFZzswr2:bokhgc2)* status: 5 of 7 files reviewed, 1 unresolved discussion (waiting on @Nateowami and @RaymondLuong3)

___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts` line 962 at r7](https://reviewable.io/reviews/sillsdev/web-xforge/1765#gh-1170392645:-NTLqbPSBdzeKiiZrZ8z:b-j9y3ak) ([raw file](https://github.com/sillsdev/web-xforge/blob/afbce02ccb39793ed601fc9d536dc9b7c5c19405/src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor.component.ts#L962)):*
<details><summary><i>Previously, Nateowami (Nathaniel Paulus) wrote…</i></summary><blockquote>

Are there not events we can subscribe to avoid this?
</blockquote></details>

Chrome are working on some APIs but they aren't supported/accepted across all browsers. I had found a library at one point to try and determine when a keyboard is open/closed. Unfortunately it requires a delay of around 2 seconds to figure it out. Part of the complexity we have is the existing resize event to grow/shrink the editor due to the performance issues we hit with letting flexbox figure it out. This event gets triggered as well when keyboards show and hide.

I'm sure we can look to improve this as time goes on and we concentrate more on our mobile experience across the entire app.

**Comment by Nateowami** [comment]
:facepalm: Yes, I was on an an older commit c88b337e. Somehow when I tried to check out the latest I messed it up and created a new branch instead. Sorry to waste your time. Both issues I pointed out are working fine.

**Pull request merged**